binders/kafka-binder/spring-cloud-stream-binder-kafka-reactive/src/main/java/org/springframework/cloud/stream/binder/reactorkafka/ReactorKafkaBinder.java

```diff
@@ -316,35 +323,40 @@ private static class ReactorMessageHandler extends AbstractMessageHandler implem
 
 		private final SenderOptions<Object, Object> senderOptions;
 
+		@Nullable
+		private final FluxMessageChannel results;
+
 		private volatile KafkaSender<Object, Object> sender;
 
 		private volatile boolean running;
 
 		ReactorMessageHandler(SenderOptions<Object, Object> opts, RecordMessageConverter converter,
-				String topic) {
+				String topic, @Nullable FluxMessageChannel results) {
 
 			this.senderOptions = opts;
 			this.converter = converter;
 			this.topic = topic;
+			this.results = results;
 		}
 
 		@Override
 		protected void handleMessageInternal(Message<?> message) {
-			Object sendResultHeader = message.getHeaders().get("sendResult");
-			Sinks.One<RecordMetadata> sink = Sinks.one();
-			if (sendResultHeader instanceof AtomicReference result) {
-				result.set(sink.asMono());
-			}
 			if (this.sender != null) {
-				UUID uuid = UUID.randomUUID();
+				Object correlation = message.getHeaders().get(IntegrationMessageHeaderAccessor.CORRELATION_ID);
+				if (correlation == null) {
+					correlation = UUID.randomUUID();
+				}
 				@SuppressWarnings("unchecked")
-				SenderRecord<Object, Object, UUID> sr = SenderRecord.create(
-						(ProducerRecord<Object, Object>) converter.fromMessage(message, topic), uuid);
-				Flux<SenderResult<UUID>> result = sender.send(Flux.just(sr));
-				result.subscribe(res -> sink.emitValue(res.recordMetadata(), null));
-			}
-			else {
-				sink.emitError(new IllegalStateException("Handler is not running"), null);
+				SenderRecord<Object, Object, Object> sr = SenderRecord.create(
+						(ProducerRecord<Object, Object>) converter.fromMessage(message, topic), correlation);
+				Flux<SenderResult<Object>> result = sender.send(Flux.just(sr));
+				result.subscribe(res -> {
```

Comment:
This is better to have like this:
```
				if (this.results != null) {
					this.results.subscribeTo(
							result.map(res ->
									MessageBuilder.withPayload(res)
											.copyHeaders(message.getHeaders())
											.build()));
				}
				else {
					result.subscribe();
				}
```
So, the `FluxMessageChannel` would subscribe to that `Flux` in its own manner.

Reply:
But wouldn't that cause a new subscription for each send? How would it be disposed of?

Reply:
There is nothing wrong with new subscription for each send.
Essentially that is exactly what you do here with this `result.subscribe()`.
Since this is limited `Flux` from `sender.send()` result, then it disposes itself after completion.

However I see the problem in the `FluxMessageChannel`: we always pile `Disposable`s into a `Disposable.Composite` and clean it up only on `FluxMessageChannel.destroy()`. 
So, this is a bug leading to memory leak.

Either way: do we talk about the same matter or you mean something else about them?

Reply:
Yes, that's what I meant. But what is wrong with using send here? It works as expected by emitting via the sink.

Reply:
Yeah... Sure! Never mind. Technically it is really the same.
Unless we don't have a memory leak with your current solution 😉